packages/compiler-cli/src/ngtsc/typecheck/extended/checks/uninvoked_track_function/index.ts

```diff
@@ -0,0 +1,81 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.dev/license
+ */
+
+import {
+  AST,
+  ASTWithSource,
+  Call,
+  PropertyRead,
+  SafeCall,
+  SafePropertyRead,
+  TmplAstForLoopBlock,
+  TmplAstNode,
+} from '@angular/compiler';
+import ts from 'typescript';
+
+import {ErrorCode, ExtendedTemplateDiagnosticName} from '../../../../diagnostics';
+import {NgTemplateDiagnostic, SymbolKind} from '../../../api';
+import {TemplateCheckFactory, TemplateCheckWithVisitor, TemplateContext} from '../../api';
+
+/**
+ * Ensures that track functions in @for loops are invoked.
+ */
+class UninvokedTrackFunctionCheck extends TemplateCheckWithVisitor<ErrorCode.UNINVOKED_TRACK_FUNCTION> {
+  override code = ErrorCode.UNINVOKED_TRACK_FUNCTION as const;
+
+  override visitNode(
+    ctx: TemplateContext<ErrorCode.UNINVOKED_TRACK_FUNCTION>,
+    component: ts.ClassDeclaration,
+    node: TmplAstNode | AST,
+  ): NgTemplateDiagnostic<ErrorCode.UNINVOKED_TRACK_FUNCTION>[] {
+    if (!(node instanceof TmplAstForLoopBlock) || !node.trackBy) {
+      return [];
+    }
+
+    if (node.trackBy.ast instanceof Call || node.trackBy.ast instanceof SafeCall) {
+      // If the method is called, skip it.
+      return [];
+    }
+
+    if (
+      !(node.trackBy.ast instanceof PropertyRead) &&
+      !(node.trackBy.ast instanceof SafePropertyRead)
+    ) {
+      // If the expression is not a property read, skip it.
+      return [];
+    }
+
+    const symbol = ctx.templateTypeChecker.getSymbolOfNode(node.trackBy.ast, component);
+
+    if (symbol !== null && symbol.kind === SymbolKind.Expression) {
+      if (symbol.tsType.getCallSignatures()?.length > 0) {
```

Comment:
```suggestion
    if (symbol?.kind === SymbolKind.Expression && symbol.tsType.getCallSignatures()?.length > 0) {
```

nit: We could make it a single condition.